Automattic/jetpack
# Widget Visibility: Restrict to top level blocks

Fixes #20057

#### Changes proposed in this Pull Request (TLDR):
* Reduce the scope of the Widget Visibility feature: No longer support recursive filtering and completely remove the `render_block()` call from `filter_widget()`. Users are only allowed to set visibility rules on the top level block of each widget.

#### Changes proposed in this Pull Request:
* In #20731, we added "Widget Visibility", an enhancement to gutenberg block widgets which allow users to apply rules that decide when blocks inside widgets display.
* When this was deployed to wpcom, we saw many crashes happening in user configurations that were not happening in a stock configuration. For example, one culprit was wpcom's infinite scroll homepage plugin.
* There is more detail about these crashes here: 
  * https://github.com/Automattic/wp-calypso/issues/56526 
  * https://github.com/Automattic/wp-calypso/issues/56546
* In all known cases, crashes were related to `filter_widget()` running `render_block()`. Before Widget Visibility was implemented, `filter_widget()` only returned its argument unchanged or returned false. With Widget Visibility, sometimes it would change the content of the block passed to it by re-rendering the block, and return that.
  * Re-rendering blocks was needed to provide support for nested rules. For example, if a widget was a column block, and the left column had visibility rules XYZ, and the right column had visibility rules ABC, we would rerender the entire column block so we could apply the visibility rules to children of the widget, instead of only the widget itself.
  * `filter_widget()` running `render_block()` is risky because so many plugins modify rendering behavior. For example, we saw an infinite loop bug where `render_block()` would need to seek out sidebar settings, which would end up getting a list of widgets and calling `filter_widget()`, which would call `render_block()`, which called `filter_widget()`.. etc.
* I was able to prevent infinite loops by using a static flag on the `Jetpack_Widget_Conditions` class, but there was at least one render related bug I couldn't fix (wp-calypso#56546).
* My solution is to reduce the scope of the Widget Visibility feature: No longer support recursive filtering and completely remove the `render_block()` call from `filter_widget()`.
  * Now `filter_widget()` works the same as before: It only ever returns false or returns its argument unchanged.
  * Scope of potential bugs is drastically reduced.
  * Now users may only set visibility settings on the outermost block of each widget.



#### Jetpack product discussion

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* `jetpack docker up`
* Make sure WP is installed and you can visit /wp-admin. You may need to use `jetpack docker install`, and possibly edit one of the env files such that your password isn't easily guessable if you decide to use a tunneling solution in the future.
* In another terminal, `jetpack watch plugins/jetpack`
  * You might also need to `jetpack build plugins/jetpack` first.
* Make sure this setting is turned on in jetpack settings.
![2021-08-17_17-46](https://user-images.githubusercontent.com/937354/129810581-2a3911d3-7670-4843-a08d-5707813ef5da.png)
* Visit http://localhost/wp-admin/widgets.php
* Add a block based widget here
* In the outermost block of any widget, go to the advanced section of the gutenberg sidebar

You should see the following in the sidebar, which lets you set up rules:
![2021-09-28_13-17](https://user-images.githubusercontent.com/937354/135143305-2a7637fc-2d8b-4b0a-86ea-29ef14982621.png)

When visiting the frontend of the site, those widgets should follow those rules.

* If you select a non-outermost block, you will see a message asking you to select the outermost block. For example, add a columns block, then add a paragraph block inside one of the columns, and attempt to set visibility rules on that paragraph block.

![2021-09-28_13-18](https://user-images.githubusercontent.com/937354/135143568-fb02e480-123b-469c-b66d-dbe766f11367.png)

#### Testing instructions (wpcom)

We need to make a custom patch, since https://github.com/Automattic/jetpack/issues/20057 is merged in JP but reverted on wpcom.

#### Testing instructions (Unit Tests)

* `jetpack docker phpunit -- --filter=Widget_Conditions`
  * In order to get the tests working, I had to: `jetpack docker sh`, then `cd /tmp/wordpress-develop/tests/phpunit/data`, then `svn up -r44722`, then `cd ../includes/` , then again `svn up -r44722`. This was done a few weeks ago and I am no longer certain this is working with new docker images.
